flang-rt/lib/runtime/tools.cpp

```diff
@@ -158,13 +196,25 @@ RT_API_ATTRS void ShallowCopy(const Descriptor &to, const Descriptor &from,
       std::memcpy(to.OffsetElement(), from.OffsetElement(),
           to.Elements() * to.ElementBytes());
     } else {
-      ShallowCopyDiscontiguousToContiguous(to, from);
+      if (to.rank() == 1 && from.rank() == 1) {
```

Comment:
It should be okay to start with rank-1 recognition here, but it is probably better to call different versions from the compiler specialized for the particular rank.

In other words, we may have `ShallowCopy` versions for each trivial data type per each constant rank.

Reply:
That's quite a lot of code duplication, no? I tried to avoid that, hence the current implementation being what it is. But if you think that's not as important a concern then I can try to implement it that way of course.

Reply:
Most of the code should go into a template.  Yes, there will be multiple instantiations, but it is not a big deal.  Yes, it will affect the size of flang-rt, but I do not think the effect will be noticeable on top of all the matmul instances.

As I said, it looks okay to me as-is to start with.